python/ray/serve/batching.py

```diff
@@ -183,34 +191,51 @@ async def _consume_func_generator(
                     elif result in [StopIteration, StopAsyncIteration]:
                         # User's code returned sentinel. No values left
                         # for caller. Terminate iteration for caller.
-                        future.set_exception(StopAsyncIteration)
+                        _set_exception_if_not_done(future, StopAsyncIteration)
                         next_futures.append(FINISHED_TOKEN)
                     else:
                         next_future = get_or_create_event_loop().create_future()
-                        # If the client has disconnected, the future is canceled.
-                        # We should only set result when the connection is still live.
-                        if not future.cancelled():
-                            future.set_result(_GeneratorResult(result, next_future))
+                        _set_result_if_not_done(
+                            future, _GeneratorResult(result, next_future)
+                        )
                         next_futures.append(next_future)
                 futures = next_futures
 
             for future in futures:
                 if future is not FINISHED_TOKEN:
-                    future.set_exception(StopAsyncIteration)
+                    _set_exception_if_not_done(future, StopAsyncIteration)
         except Exception as e:
             for future in futures:
                 if future is not FINISHED_TOKEN:
-                    future.set_exception(e)
+                    _set_exception_if_not_done(future, e)
 
-    async def _process_batches(self, func: Callable) -> None:
+    async def _process_batches_forever(self, func: Callable) -> None:
```

Comment:
nit: I don't think "forever" qualification adds much in this case

Reply:
I changed it back to _process_batches.